### PR TITLE
Implements foundations of aggregation framework.

### PR DESCRIPTION
**Pull request by vladlosev**
@bigthyme @parkr This is a foundation of aggregation framework. Only `$group` stage is currently supported, and only `$sum` operation in it, but it's something we can build on.

### CODE REVIEW

**Comment by parkr** [comment]
NICE!

File: lib/commands/aggregate.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  return {documents: [error]};
+}
+
+// We are implementing aggregation groups via objects. Objects in JavaScript
+// can only have strings as keys, but MongoDB aggregations can be based on all
+// kinds of JavaScript types.  In order to distinguish for example, between a
+// number 2 and a string '2', we need some extra logic which this function
+// provides.
+function buildValueKey(value) {
+  if (utils.isObjectId(value)) {
```
Comment (bigthyme) [inline lib/commands/aggregate.js]: would this be better suited for a switch statement?

File: lib/commands/aggregate.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    return 'objectid:' + value;
+  } else if (_.isNumber(value)) {
+    return 'number:' + value;
+  } else if (_.isString(value)) {
+    return 'string:' + value;
+  } else if (_.isDate(value)) {
+    return 'date:' + value.toJSON();
+  } else if (value === null || value === undefined) {
```
Comment (bigthyme) [inline lib/commands/aggregate.js]: if you're using `_` for validation, you should also use it here `_.isNull` and `_.isUndefined`

File: lib/commands/aggregate.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    return 'null:';
```
Comment (bigthyme) [inline lib/commands/aggregate.js]: is this null supposed to have a colon?

File: lib/commands/aggregate.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// Retrieves the a value for aggregating, based on the field expression
+// supplied in an aggregation stage.
+// TODO(vladlosev): Extend this to handle different functions defined in the
+// aggregation framework and variables.
+function getFieldExpressionValue(doc, expression) {
+  if (/^[$]/.test(expression)) {
+    var value = wrapElementForAccess(doc, expression.substring(1)).getValue();
+    if (value === undefined) {
```
Comment (bigthyme) [inline lib/commands/aggregate.js]: same here with the `_.isUndefined`

File: lib/commands/aggregate.js (hunk 3 above)
Comment (vladlosev) [inline lib/commands/aggregate.js]: The idea is for `:` to separate the type from the value. But once you know it's a `null`, there is nothing more to say. We may revisit this later, but it works so far.

File: lib/commands/aggregate.js (hunk 2 above)
Comment (vladlosev) [inline lib/commands/aggregate.js]: Yay for consistency! Done.

File: lib/commands/aggregate.js (hunk 1 above)
Comment (vladlosev) [inline lib/commands/aggregate.js]: Because the underlying checks are not scalar comparisons, there will always be a bunch of `if` statements hiding somewhere around.

File: lib/commands/aggregate.js (hunk 4 above)
Comment (vladlosev) [inline lib/commands/aggregate.js]: Done.

**Comment by bigthyme** [comment]
LGTM!

**Comment by vladlosev** [comment]
I decided to add a couple more tests for some edge cases and that uncovered some bugs. The new commits fix them and do a small re-factoring to ease subsequent adding of more operators. PTAL.
